### PR TITLE
Implement basic skeleton & networking

### PR DESCRIPTION
**Pull request by AndrejMitrovic**
I've stripped out everything from the draft project by @Geod24 and left in the networking. It's basic stuff, but it works, and we can get started implementing other things like interacting with SCP, etc.

I can demonstrate how multiple nodes connect with each other in person.

Edit: I wish to make everything unit-testable, perhaps with the framework @Geod24 previously implemented.

### CODE REVIEW

File: doc/config.example.yaml [hunk 1]
```diff
@@ -0,0 +1,82 @@
+################################################################################
+## This is a sample, documented configuration file for <PROJECT-NAME>
```
Comment (Geod24) [inline doc/config.example.yaml]: <CURRENT-YEAR>

File: source/agora/app.d [hunk 2]
```diff
@@ -12,12 +20,53 @@
...
+import vibe.web.rest;
 
-void main()
+import std.digest.sha;
+import std.stdio;
+import core.stdc.time;
+import core.time;
+
+/// todo: work around this for 'dub test'
+version (unittest) void main () { } else:
```
Comment (Geod24) [inline source/agora/app.d:31]: Shouldn't be necessary if you specify `mainSourceFile` in `dub.json`

File: source/agora/common/API.d [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    Contains the REST API interface.
+
+    Copyright:
+        Copyright (c) 2019 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+
```
Comment (Geod24) [inline source/agora/common/API.d]: W H I T E S P A C E

File: source/agora/common/Hash.d [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    static if (is(T : E[], E))
+        const bs = ByteSlice.make(arg);
+    else static if (is(typeof(arg.serialize(sink))))
+    {
+        arg.serialize(sink);
+        const bs = ByteSlice.make(buffer);
+    }
+    else
+        const bs = ByteSlice.make(XDRToOpaque(arg));
+    return sha256(bs);
```
Comment (Geod24) [inline source/agora/common/Hash.d]: Is that ?... I mean ? I'm scared.

File: source/agora/common/Set.d [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+/// simplified Set() code with some convenience methods,
+/// could use a drop-in implementation later.
+struct Set (T)
+{
+    ///
+    bool[][T] _set;
+    alias _set this;
+
+    /// Get and remove one element from the set
+    public T stealOne ( ) ( )
```
Comment (Geod24) [inline source/agora/common/Set.d]: Why is this a template ?

File: source/agora/common/Hash.d (hunk 4 above)
Comment (AndrejMitrovic) [reply source/agora/common/Hash.d]: This seems to be unused leftover code from _that other repo of yours_. I used it during testing here and there, I'll remove it.

**Comment by AndrejMitrovic** [comment]
Update:

- Fixed documented config to mention Agora.
- Fixed extra blank line
- Added config verification which should be called both after parsing, and after creating a manual config for unittests
- Made array fields in the config immutable (see next line)
- Added shared aliasing tests to make sure these configs can always be used with the LocalRest library
- Removed unecessary key_pair field which can just be part of the Config file instead now that it's immutable
- Changed retry-delay from integer to floating-point. This allows setting it to a few milliseconds which is useful in order to speed up unittests.
- Fixed uneccessary "() ()" template
- Removed unused Hash module
- Fixed a node from trying to connect to itself during discovery
- Fixed fiber sleep calls to consistently use the configured sleep time
- Moved creation of the initial task directly into the Node class. This allows the node to auto-connect in unittests without having to manually run its node.start() method.

And:
- Added a commit implementing a very basic network discovery test-case. Much more will be added, this was just a proof-of-concept. But the test already uncovered a few bugs that are now fixed.

**Comment by AndrejMitrovic** [comment]
> @AndrejMitrovic AndrejMitrovic force-pushed the AndrejMitrovic:basic-networking  branch from ace94d4 to 04b9a1d 1 minute ago

Unfortunately this shows the wrong diff because I force-pushed the wrong changes at first.

File: doc/config.example.yaml [hunk 6]
```diff
@@ -0,0 +1,81 @@
+################################################################################
+## This is a sample, documented configuration file for Agora in YAML format
+##
+## It is not intended for usage on any network
+##
+## A node has two interface: a working and an administrative one
```
Comment (Geod24) [inline doc/config.example.yaml]:
```suggestion
## A node has two interface: one for network communication and one for administration
```

File: doc/config.example.yaml [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+## The public one is used when communicating with the network.
```
Comment (Geod24) [inline doc/config.example.yaml]: And then rephrase this as:
```
The one communicating with the network is public, while the administrative is unsecured, disabled by default, and should not be exposed to the outside world.
```

File: doc/config.example.yaml [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+## The administrative one is unsecured and should not be exposed to the outside
+## world. It is disabled by default for security purpose.
+################################################################################
+
+################################################################################
+##                             Network interface                              ##
+################################################################################
+node:
+  # The server can operate in two modes: watcher and validator
```
Comment (Geod24) [inline doc/config.example.yaml]: `s/watcher/full node/g`

File: doc/config.example.yaml [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+################################################################################
+##                               Node discovery                               ##
+##                                                                            ##
+## When the network first starts, we need to connect to some peers to learn   ##
+## the topology and find a safe intersection to listen to, and, if we are     ##
+## a validator, to insert ourselves.                                          ##
+################################################################################
+network:
+  # Supported value: domain name, IPv4, IPv6
+  - http://0.0.0.0:2828
```
Comment (Geod24) [inline doc/config.example.yaml]:
```suggestion
  - http://192.168.1.42:2828
  - http://192.168.0.44:2828
```

File: source/agora/app.d [hunk 10]
```diff
@@ -12,12 +20,38 @@
...
+
+/// Application entry point
+public int main (string[] args)
```
Comment (Geod24) [inline source/agora/app.d]: `private`

File: source/agora/app.d [hunk 11]
```diff
@@ -12,12 +20,38 @@
...
 {
-	// TODO:
-    // - Config parsing
-    // - Network interface setup
-    // - Event loop
-    // - End world hunger
+    CommandLine cmdln;
+    if (!parseCommandLine(cmdln, args))
+        return 0;  // help shown, early exit
```
Comment (Geod24) [inline source/agora/app.d]:
```suggestion
        return 1;  // help shown, early exit
```

File: source/agora/common/Config.d [hunk 12]
```diff
@@ -0,0 +1,310 @@
...
+    string config_path = "config.yaml";
+
+    /// check state of config file and exit early
+    bool config_check;
+}
+
+/// Main config
+public struct Config
+{
+    static assert(!hasUnsharedAliasing!(typeof(this)));
```
Comment (Geod24) [inline source/agora/common/Config.d]: Add a problem explaining the comment why

File: source/agora/node/Node.d [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+
+module agora.node.Node;
+
+import agora.common.API;
+import agora.common.Config;
+import agora.common.crypto.Key;
+import agora.common.Version;
+import agora.node.Network;
+
+import vibe.d;
```
Comment (Geod24) [inline source/agora/node/Node.d:21]: Nope!

File: source/agora/node/NodeTest.d [hunk 14]
```diff
@@ -0,0 +1,364 @@
...
+
+    //case NetworkTopology.Balanced:
+    //    assert(nodes.length == 4, "Balanced topology requires 4 nodes");
+    //    assert(0, "Not implemented");
+
+    //case NetworkTopology.Tiered:
+    //    assert(nodes.length == 10, "Tiered topology requires 10 nodes");
+    //    assert(0, "Not implemented");
+
+    //case NetworkTopology.Cyclic:
```
Comment (Geod24) [inline source/agora/node/NodeTest.d]: Remove dead code

File: source/agora/node/NodeTest.d [hunk 15]
```diff
@@ -0,0 +1,364 @@
...
+    //    assert(nodes.length == 7, "SinglePoF topology requires 7 nodes");
+    //    assert(0, "Not implemented");
+    }
+}
+
+///
+unittest
+{
+    import core.thread;
+    import std.stdio;
```
Comment (Geod24) [inline source/agora/node/NodeTest.d:316]: Why the imports ?

File: source/agora/node/NodeTest.d [hunk 16]
```diff
@@ -0,0 +1,364 @@
...
+    Copyright:
+        Copyright (c) 2019 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.node.NodeTest;
```
Comment (Geod24) [inline source/agora/node/NodeTest.d:14]: I think this should live in a separate package (`agora.test` maybe ?)

File: source/agora/node/NodeTest.d [hunk 17]
```diff
@@ -0,0 +1,364 @@
...
+
+        enforce(fully_discovered.length == this.registry.length,
+            format("Got %s/%s discovered nodes. Missing nodes: %s",
+                fully_discovered.length,
+                this.registry.length,
+                this.registry.byKey.filter!(a => !!(a in fully_discovered))));
+
+        // todo: trace logging during unittesting?
+        // writefln("Discovered %s/%s nodes.", fully_discovered.length,
+        //     this.registry.length);
```
Comment (Geod24) [inline source/agora/node/NodeTest.d]: Remove dead code

File: source/agora/node/NodeTest.d [hunk 18]
```diff
@@ -0,0 +1,364 @@
...
+
+    ///
+    public override API[] getAllAPIs ( )
+    {
+        return this.registry.byValue.array;
+    }
+}
+
+///
+class TestNodeRegistry : TestRegistry
```
Comment (Geod24) [inline source/agora/node/NodeTest.d:120]: For the moment let's keep it as it, but the separation was done so that you could have different registry for different tests (different registry == different node behavior)

File: source/agora/node/Network.d [hunk 19]
```diff
@@ -0,0 +1,240 @@
...
+import agora.common.crypto.Key;
+import agora.common.Config;
+import agora.common.Data;
+import agora.common.Set;
+import agora.node.Client;
+
+import libsodium.crypto_sign_ed25519;
+import libsodium.randombytes;
+
+import vibe.d;
```
Comment (Geod24) [inline source/agora/node/Network.d]: Nope, this is fine in test code but not here

File: source/agora/node/Client.d [hunk 20]
```diff
@@ -0,0 +1,193 @@
...
+module agora.node.Client;
+
+import agora.common.API;
+import agora.common.Config;
+import agora.common.crypto.Key;
+import agora.common.Data;
+import agora.common.Set;
+import agora.common.Version;
+
+import libsodium.crypto_sign_ed25519;
```
Comment (Geod24) [inline source/agora/node/Client.d]: Leftover ?

File: source/agora/common/Version.d [hunk 21]
```diff
@@ -0,0 +1,37 @@
...
+    Copyright:
+        Copyright (c) 2019 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.common.Version;
```
Comment (Geod24) [inline source/agora/common/Version.d]: Why this module ?

File: source/agora/node/Network.d (hunk 19 above)
Comment (AndrejMitrovic) [reply source/agora/node/Network.d]: You mean using the package module? Why is it not ok?

File: source/agora/common/Version.d (hunk 21 above)
Comment (AndrejMitrovic) [reply source/agora/common/Version.d]: Useless actually, can remove it for now.

File: source/agora/node/NodeTest.d (hunk 18 above)
Comment (AndrejMitrovic) [reply source/agora/node/NodeTest.d:120]: Btw, how exactly do we plan on simulating a "bad" node? In theory you can only interact with it through its publicly exposed API.

Comment (AndrejMitrovic) [reply source/agora/node/NodeTest.d:120]: I'm guessing through subclassing the Node class somehow?

File: source/agora/common/Config.d (hunk 12 above)
Comment (AndrejMitrovic) [reply source/agora/common/Config.d]: _because immutable sucks_ :PP

File: source/agora/app.d (hunk 10 above)
Comment (AndrejMitrovic) [reply source/agora/app.d]: I can change it if you want but I don't understand why or what for?

File: doc/config.example.yaml (hunk 6 above)
Comment (AndrejMitrovic) [reply doc/config.example.yaml]: It's really cute when you review your own text. :P

File: source/agora/app.d (hunk 10 above)
Comment (Geod24) [reply source/agora/app.d]: Just a matter of style I guess. Why would you have it public ? No one should call it.

File: source/agora/node/NodeTest.d (hunk 18 above)
Comment (Geod24) [reply source/agora/node/NodeTest.d:120]: Yes

File: source/agora/app.d (hunk 10 above)
Comment (AndrejMitrovic) [reply source/agora/app.d]: Conceptually it's the entry point of the application. In practice the real main function is in DRuntime.. anyway I'll make it private.

File: source/agora/node/Network.d (hunk 19 above)
Comment (Geod24) [reply source/agora/node/Network.d]: Too many dependencies being pulled in

File: source/agora/node/NodeTest.d (hunk 18 above)
Comment (Geod24) [reply source/agora/node/NodeTest.d:120]: You subclass the node class and implement the behavior. Then in `registry`, you give it a name, e.g. `byzantine` or `dead` and return the appropriate node.

File: source/agora/node/Network.d (hunk 19 above)
Comment (AndrejMitrovic) [reply source/agora/node/Network.d]: Good point.

File: source/agora/node/NodeTest.d (hunk 18 above)
Comment (AndrejMitrovic) [reply source/agora/node/NodeTest.d:120]: Ah, I see now why you have that name string passed as a parameter. Got it.

File: source/agora/app.d (hunk 11 above)
Comment (Geod24) [reply source/agora/app.d]: ^ After reading the code, forget about this

Comment (AndrejMitrovic) [reply source/agora/app.d]: I could make it an enum to make it clear?

Comment (Geod24) [reply source/agora/app.d]: Some command line tools returns success on `--help` (e.g. `git`), some don't.
Let's follow git here

File: source/agora/app.d (hunk 2 above)
Comment (AndrejMitrovic) [reply source/agora/app.d:31]: I tried doing this:

```

"configurations": [
    {
        "name": "agora",
        "targetName": "agora",
        "targetType": "executable",
        "mainSourceFile": "source/agora/app.d",
    },
    {
        "name": "unittest",
        "sourceFiles": [
            "source/scpp/build/DSizeChecks.o",
            "source/scpp/build/DLayoutChecks.o"
        ],
        "excludedSourceFiles": [ "source/agora/app.d" ]
    }
],
```

But then I get a linker error with `dub test`.

**Comment by AndrejMitrovic** [comment]
~I fixed most stuff except the `dub test` target which still fails. Why does dub have to be so difficult..~ Edit: fixed

**Comment by AndrejMitrovic** [comment]
There is also some missing documentation. Can be added soon or later after merging if I don't make it today.

**Comment by AndrejMitrovic** [comment]
Ok, managed to fix the dub test target.

**Comment by AndrejMitrovic** [comment]
I also renamed `Client` to `RemoteNode` to make it a bit clearer.

**Comment by AndrejMitrovic** [comment]
> Ok, managed to fix the dub test target.

~Nope..~

**Comment by AndrejMitrovic** [comment]
So the reason why I was confused that my tests didn't always build/link is because in the first commit `dub test` fails, but in the second one it doesn't. It's really weird. Anyway it doesn't matter, second one works fine.

**Review by Geod24 — APPROVED** [review]
